### PR TITLE
Add extra configurability to `parallelizeAllLike`

### LINKED ISSUE #1828: scheduler_utils::parallelizeAllLike is changing parallelization that it should not change
### 🐛 Describe the bug

I am trying to schedule a transpose kernel manually using the following code:

```C++


// Copied from Shiming's PR
struct MatmulSelector : public MaxInfoSpanningTree::Selector {
 public:
  explicit MatmulSelector(std::unordered_set<TensorView*> selected_tvs)
      : selected_tvs_(selected_tvs) {}

  explicit MatmulSelector() : propagate_all_(true) {}

  bool allowC2P(TensorView* from, TensorView* to) final {
    return propagate_all_ || selected_tvs_.count(to);
  }
  bool allowP2C(TensorView* from, TensorView* to) final {
    return propagate_all_ || selected_tvs_.count(to);
  }
  bool allowSibling(TensorView* from, TensorView* to) final {
    return propagate_all_ || selected_tvs_.count(to);
  }

 private:
  std::unordered_set<TensorView*> selected_tvs_;
  bool propagate_all_ = false;
};

// Copied from Shiming's PR, modified
void propagateFrom(
    TensorView* from_tv,
    int pos,
    std::unordered_set<TensorView*> included_tvs,
    bool propagate_transformation,
    bool propagate_parallel_type,
    bool propagate_inline_position) {
  MatmulSelector selector(included_tvs);
  MaxRootDomainInfoSpanningTree spanning_tree(from_tv, &selector);

  if (propagate_transformation) {
    TransformPropagator propagator(from_tv, pos);
    spanning_tree.traverse(&propagator);
  }

  if (propagate_parallel_type) {
    scheduler_utils::parallelizeAllLike(
        from_tv, {included_tvs.begin(), included_tvs.end()});
  }

  if (propagate_inline_position) {
    InlinePropagator inline_propagator(from_tv, pos, included_tvs);
    spanning_tree.traverse(&inline_propagator);
  }
}

TEST_F(NVFuserTest, Transpose2) {
  Fusion fusion;
  FusionGuard fg(&fusion);

  auto tv0 = makeSymbolicTensor(3);
  fusion.addInput(tv0);
  auto tv1 = sin(tv0);
  auto tv2 = transpose(tv1, 1, 2);
  auto tv3 = cos(tv2);
  fusion.addOutput(tv3);

  auto tv4 = tv0->cacheAfter();
  auto tv5 = tv3->cacheBefore();

  tv1->setMemoryType(MemoryType::Shared);

  tv1->split(1, 32);
  tv1->reorder({{2, -1}});
  tv1->split(2, 32);
  tv1->reorder({{3, -1}});

  tv1->axis(0)->parallelize(ParallelType::BIDz);
  tv1->axis(1)->parallelize(ParallelType::BIDy);
  tv1->axis(2)->parallelize(ParallelType::BIDx);
  tv1->axis(3)->parallelize(ParallelType::TIDy);
  tv1->axis(4)->parallelize(ParallelType::TIDx);

  tv2->split(1, 32);
  tv2->reorder({{2, -1}});
  tv2->split(2, 32);
  tv2->reorder({{3, -1}});
  tv2->reorder({{1, 2}});

  tv2->axis(0)->parallelize(ParallelType::BIDz);
  tv2->axis(1)->parallelize(ParallelType::BIDy);
  tv2->axis(2)->parallelize(ParallelType::BIDx);
  tv2->axis(3)->parallelize(ParallelType::TIDy);
  tv2->axis(4)->parallelize(ParallelType::TIDx);

  fusion.print();

  propagateFrom(tv1, -1, {tv1, tv0, tv4}, true, true, false);
  propagateFrom(tv1, 3, {tv1, tv0, tv4}, false, false, true);

  fusion.print();

  propagateFrom(tv2, -1, {tv2, tv3, tv5}, true, true, false);
  propagateFrom(tv2, 3, {tv2, tv3, tv5}, false, false, true);

  fusion.printKernel();
}
```

And I am seeing strange behavior: At the first print, `tv1` and `tv2` has:
```
T1_s[ iblockIdx.z3{i1}, iblockIdx.y18{( ceilDiv(i2, 32) )}, iblockIdx.x20{( ceilDiv(i3, 32) )}, ithreadIdx.y19{32}, ithreadIdx.x21{32} ]
 root domain : (iblockIdx.z3{i1},iS4{i2},iS5{i3})
  Split: iS4{i2} by factor 32 -> iblockIdx.y18{( ceilDiv(i2, 32) )}, ithreadIdx.y19{32}, start offset: 0, stop offset: 0
  Split: iS5{i3} by factor 32 -> iblockIdx.x20{( ceilDiv(i3, 32) )}, ithreadIdx.x21{32}, start offset: 0, stop offset: 0
T2_l[ iblockIdx.z6{i1}, iblockIdx.y24{( ceilDiv(i2, 32) )}, iblockIdx.x22{( ceilDiv(i3, 32) )}, ithreadIdx.y23{32}, ithreadIdx.x25{32} ]
 root domain : (iblockIdx.z6{i1},iS7{i3},iS8{i2})
  Split: iS8{i2} by factor 32 -> iblockIdx.y24{( ceilDiv(i2, 32) )}, ithreadIdx.x25{32}, start offset: 0, stop offset: 0
  Split: iS7{i3} by factor 32 -> iblockIdx.x22{( ceilDiv(i3, 32) )}, ithreadIdx.y23{32}, start offset: 0, stop offset: 0
```
which is expected. But at the second print, they become:
```
T1_s[ iblockIdx.z3{i1}, iblockIdx.y18{( ceilDiv(i2, 32) )}, iblockIdx.x20{( ceilDiv(i3, 32) )}, ithreadIdx.y19{32}, ithreadIdx.x21{32} ] ca_pos( 3 ) produce_pos( 3)
 root domain : (iblockIdx.z3{i1},iS4{i2},iS5{i3})
  Split: iS4{i2} by factor 32 -> iblockIdx.y18{( ceilDiv(i2, 32) )}, ithreadIdx.y19{32}, start offset: 0, stop offset: 0
  Split: iS5{i3} by factor 32 -> iblockIdx.x20{( ceilDiv(i3, 32) )}, ithreadIdx.x21{32}, start offset: 0, stop offset: 0
T2_l[ iblockIdx.z6{i1}, iblockIdx.y24{( ceilDiv(i2, 32) )}, iblockIdx.x22{( ceilDiv(i3, 32) )}, ithreadIdx.x23{32}, ithreadIdx.y25{32} ] produce_pos( 3)
 root domain : (iblockIdx.z6{i1},iS7{i3},iS8{i2})
  Split: iS8{i2} by factor 32 -> iblockIdx.y24{( ceilDiv(i2, 32) )}, ithreadIdx.y25{32}, start offset: 0, stop offset: 0
  Split: iS7{i3} by factor 32 -> iblockIdx.x22{( ceilDiv(i3, 32) )}, ithreadIdx.x23{32}, start offset: 0, stop offset: 0
```
This is not what I expect, because `tv2` is not in the selected set of tensors that I want to change, but the parallelization of the last two dims of `tv2` is swapped.

The root cause seems to be this:
https://github.com/csarofeen/pytorch/blob/3b87896706fc98aa4d5b5c811af034cc4dddfbab/torch/csrc/jit/codegen/cuda/scheduler/utils.cpp#L197-L198

I don't understand:
- Why do we want to change the concrete mapped id's parallelization regardless of the tensor it belongs to, and
- If the dim I am currently looking at is not a broadcasting dim, isn't it already concrete? Why are we still looking for a concretely mapped dim?

### Versions

devel

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Fixes #1828 #1836

New interface:
```C++
TORCH_CUDA_CU_API void parallelizeAllLike(
    TensorView* reference_tv,
    int64_t pos = -1,
    std::unordered_set<TensorView*> selected_tvs = {}, // empty means all tvs
    const std::unordered_set<ParallelType>& selected_parallel_types =
        {}, // empty means all parallel types
    bool propagate_padding = true);
```
Also added `allTvsExcept` and `allParallelTypesExcept`.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
Making this WIP, and I will add more features as described in https://github.com/csarofeen/pytorch/issues/1836

File: torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp [hunk 1]
```diff
@@ -822,37 +821,27 @@ void schedulePointwise(Fusion* fusion, const PointwiseParams& params) {
...
-          if (id->getParallelType() == ParallelType::Vectorize) {
-            id->parallelize(ParallelType::Serial);
-          }
-        }
-      }
+    vectorize_id->parallelize(ParallelType::Vectorize);
+    scheduler_utils::parallelizeAllLike(
+        reference_tv, vectorized_tvs, {ParallelType::Vectorize});
+    if (vectorized_tvs.count(reference_tv) == 0) {
+      vectorize_id->parallelize(ParallelType::Serial);
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp]: This is just a cleanup of the code using the new functionality. There should be no behavioral change.

File: torch/csrc/jit/codegen/cuda/ir_utils.h [hunk 2]
```diff
@@ -282,6 +282,12 @@ TORCH_CUDA_CU_API std::vector<TensorView*> outputTvsOf(
 // returns all tensor views in fusion that are used between outputs and inputs.
 TORCH_CUDA_CU_API std::vector<TensorView*> allTvs(Fusion* fusion);
 
+// returns all tensor views in fusion that are used between outputs and inputs
+// except the specified set.
+TORCH_CUDA_CU_API std::unordered_set<TensorView*> allTvsExcept(
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_utils.h]: Might want to return a vector for reproducibility

File: torch/csrc/jit/codegen/cuda/scheduler/utils.h [hunk 3]
```diff
@@ -51,7 +51,24 @@ size_t mergeNonReduction(
 
 TORCH_CUDA_CU_API void parallelizeAllLike(
     TensorView* reference_tv,
-    const std::vector<TensorView*>& all_tvs);
+    int64_t pos = -1,
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/scheduler/utils.h:61]: Please add  a comment. Propagation is done only for the first `pos` domains, right?

File: torch/csrc/jit/codegen/cuda/scheduler/utils.cpp [hunk 4]
```diff
@@ -188,30 +188,54 @@ size_t mergeNonReduction(
...
   auto ca_map = ComputeAtMap(FusionGuard::getCurFusion());
 
-  for (auto id : reference_tv->domain()->domain()) {
-    ca_map.getConcreteMappedID(id, IdMappingMode::PERMISSIVE)
-        ->parallelize(id->getParallelType());
-    if (id->hasPaddingToMultipleOfWarp()) {
-      ca_map.getConcreteMappedID(id, IdMappingMode::PERMISSIVE)
-          ->padToMultipleOfWarp(id->getMaybeSizeAfterPadding());
-    }
+  auto reference_dom = reference_tv->domain()->domain();
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/scheduler/utils.cpp]: nit: const auto&

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 2 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: fixed

File: torch/csrc/jit/codegen/cuda/scheduler/utils.cpp (hunk 4 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/scheduler/utils.cpp]: fixed

File: torch/csrc/jit/codegen/cuda/scheduler/utils.h (hunk 3 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/scheduler/utils.h:61]: Added on top of `parallelizeAllLike`. And you are right, `pos` means selecting the first `pos` IDs.

**Review by naoyam — APPROVED** [review]
LGTM

**Pull request merged**